mempool/src/tx_accumulator.rs

```diff
@@ -65,35 +65,55 @@ impl DefaultTxAccumulator {
     pub fn new(target_size: usize, expected_tip: Id<GenBlock>, timestamp: BlockTimestamp) -> Self {
         Self {
             txs: Vec::new(),
-            total_size: 0,
+            txs_size: 0,
             target_size,
             done: false,
             total_fees: Amount::ZERO.into(),
             expected_tip,
             timestamp,
         }
     }
+
+    pub fn total_size(&self) -> usize {
+        Compact(self.transactions().len() as u64).encoded_size() + self.txs_size
+    }
+
+    // Calculate total size with an extra transaction of given size
+    fn total_size_with(&self, new_size: usize) -> usize {
+        Compact(self.transactions().len() as u64 + 1).encoded_size() + self.txs_size + new_size
+    }
 }
 
 impl TransactionAccumulator for DefaultTxAccumulator {
     fn add_tx(&mut self, tx: SignedTransaction, tx_fee: Fee) -> Result<(), TxAccumulatorError> {
-        if self.total_size + tx.encoded_size() <= self.target_size {
-            self.total_size += tx.encoded_size();
+        let tx_size = tx.encoded_size();
+        let total_size_with_tx = self.total_size_with(tx_size);
+
+        if total_size_with_tx <= self.target_size {
+            self.txs_size += tx_size;
             self.total_fees = (self.total_fees + tx_fee).ok_or(
                 TxAccumulatorError::FeeAccumulationError(self.total_fees, tx_fee),
             )?;
             self.txs.push(tx);
+
+            // Sanity check that total_size_with() and total_size() agree
+            assert_eq!(total_size_with_tx, self.total_size());
         } else {
-            self.done = true
+            self.done = true;
         };
+
+        // Sanity check the size tracking is accurate
+        #[cfg(test)]
+        assert_eq!(self.total_size(), self.transactions().encoded_size());
```

Comment:
I would change this to a log message starting with CRITICAL. The last thing we want is an off-by-one error crashing the node.

Reply:
Never mind. I just noticed the cfg(test). Cheers!